dace/transformation/interstate/gpu_transform_sdfg.py

```diff
@@ -192,27 +192,196 @@ def apply(self, _, sdfg: sd.SDFG):
+                    # If the tasklet is not adjacent only to scalars or it is in a GPU scope.
+                    # The latter includes NestedSDFGs that have a GPU-Device schedule but are not in a GPU kernel.
+                    if (not scalar_output or (csdfg.parent is not None
+                                              and csdfg.parent_nsdfg_node.schedule == dtypes.ScheduleType.GPU_Default)):
+                        global_code_nodes[state].append(node)
+                        gpu_scalars.update({k: None for k in scalars})
+
+        #######################################################
+        # Step 6: Modify transient data storage
 
         const_syms = xfh.constant_symbols(sdfg)
 
+        def _codenode_condition(node):
+            return ((isinstance(node, (nodes.LibraryNode, nodes.NestedSDFG))
+                     and node.schedule == dtypes.ScheduleType.GPU_Default))
+
         for state in sdfg.nodes():
             sdict = state.scope_dict()
             for node in state.nodes():
                 if isinstance(node, nodes.AccessNode) and node.desc(sdfg).transient:
                     nodedesc = node.desc(sdfg)
 
-                    # Special case: nodes that lead to dynamic map ranges must
-                    # stay on host
+                    # Special case: nodes that lead to dynamic map ranges must stay on host
                     if any(isinstance(state.memlet_path(e)[-1].dst, nodes.EntryNode) for e in state.out_edges(node)):
                         continue
 
-                    gpu_storage = [
-                        dtypes.StorageType.GPU_Global, dtypes.StorageType.GPU_Shared, dtypes.StorageType.CPU_Pinned
-                    ]
                     if sdict[node] is None and nodedesc.storage not in gpu_storage:
-                        # NOTE: the cloned arrays match too but it's the same
-                        # storage so we don't care
+
+                        # Ensure that scalars not already GPU-marked are actually used in a GPU scope.
+                        if isinstance(nodedesc, data.Scalar) and not node.data in gpu_scalars:
+                            used_in_gpu_scope = False
+                            for e in state.in_edges(node):
+                                if _codenode_condition(state.memlet_path(e)[0].src):
+                                    used_in_gpu_scope = True
+                                    break
+                            if not used_in_gpu_scope:
+                                for e in state.out_edges(node):
+                                    if _codenode_condition(state.memlet_path(e)[-1].dst):
+                                        used_in_gpu_scope = True
+                                        break
+                            if not used_in_gpu_scope:
+                                continue
+                            for e in state.all_edges(node):
+                                for node in (e.src, e.dst):
```

Comment:
shouldn't this be using memlet_path[0].src and [-1].dst?

Reply:
Since, `sdict[node] is None`, the memlet path should always consist of a single edge, right? Which means that the use of `memlet_path` above is uneeded ... I need to check again in the next PR.

Reply:
True.